src/acl.c

```diff
@@ -1243,22 +1449,47 @@ user *ACLGetUserByName(const char *name, size_t namelen) {
  *
  * If the user can access the key, ACL_OK is returned, otherwise
  * ACL_DENIED_KEY is returned. */
-int ACLCheckKey(const user *u, const char *key, int keylen) {
+int ACLCheckKey(user *u, const char *key, int keylen) {
     /* If there is no associated user, the connection can run anything. */
     if (u == NULL) return ACL_OK;
 
     /* The user can run any keys */
-    if (u->flags & USER_FLAG_ALLKEYS) return ACL_OK;
+    aclSelector *selector = ACLUserGetDefaultSelector(u);
+    if (selector->flags & SELECTOR_FLAG_ALLKEYS) return ACL_OK;
 
     listIter li;
     listNode *ln;
-    listRewind(u->patterns,&li);
+    listRewind(selector->patterns,&li);
 
     /* Test this key against every pattern. */
     while((ln = listNext(&li))) {
-        sds pattern = listNodeValue(ln);
-        size_t plen = sdslen(pattern);
-        if (stringmatchlen(pattern,plen,key,keylen,0))
+        keyPattern *pattern = listNodeValue(ln);
+        size_t plen = sdslen(pattern->pattern);
+        if (stringmatchlen(pattern->pattern,plen,key,keylen,0))
+            return ACL_OK;
+    }
+    return ACL_DENIED_KEY;
+}
+
+/* Check if the key can be accessed by the client according to
+ * the ACLs associated with the specified user.
+ *
+ * If the user can access the key, ACL_OK is returned, otherwise
+ * ACL_DENIED_KEY is returned. */
+int ACLSelectorCheckKey(aclSelector *selector, const char *key, int keylen, int flags) {
```

Comment:
I added some more docs here and took the suggestion to make them static. I also moved selectors internal to acl.c, since they shouldn't be exposed outside of the acl code. They're basically an implementation detail.